### ISSUE TITLE
Document how to assign a rarity weight of less than 1 (like 0.15). Doing #0.15.png does not work. 

### ISSUE THREAD

**Issue by purpleglow** [HashLips/hashlips_art_engine#251]
Please document how to assign a rarity weight of less than 1 , like 0.15 or 0.5. Thank you.

**Comment by edenheijer** [comment]
Good day,

As the chance of occurrence of an item is its rarity weight divided by the total rarity weight (sum of all the individual weights) times 100, I would suggest a work around by multiplying the weights by 10 to the power of the  desired decimal precision  (e.g. x10 of you would want a weight of 0.5 which would then be 5, x100 for two decimals etc etc).

The ratio remains the same and you get around working with , for example,  regular expressions to parse fractions.

**Comment by aminebidda** [comment]
> Good day,
> 
> As the chance of occurrence of an item is its rarity weight divided by the total rarity weight (sum of all the individual weights) times 100, I would suggest a work around by multiplying the weights by 10 to the power of the desired decimal precision (e.g. x10 of you would want a weight of 0.5 which would then be 5, x100 for two decimals etc etc).
> 
> The ratio remains the same and you get around working with , for example, regular expressions to parse fractions.

Can you please explain it more?

**Comment by edenheijer** [comment]
The 'weight' you assign to an individual item, after the '#', is NOT a percentage, as explained in the video tutorial linked elsewhere.  

Let's say you would like to have 0.15 % chance of occurrence of example trait blueeyes and 99.85 % of occurrence of browneyes. If the program would have supported fractions, you would want to name the files:

blueeyes#0.15.png
browneyes#99.85.png

Because the program adds all the weights together and calculates the odds by dividing the weight by the total sum we however can achieve this by naming the files:

blueeyes#15.png
browneyes#9985.png

The program first adds all the weights (15 + 9985 = 10000) and calculates the odds of occurrence by dividing the individual weight by the total sum (15 / 10000). Because we multiplied the numerator and denominator by the same factor, the ratio remains the same, in other words, ( 0.15 / 99.85 ) = (15 / 9985)

TLDR; if you want to use weights smaller than 1, multiply all weights by 10^(number of decimals of the smallest weight) 

blueeyes#15

**Comment by purpleglow** [comment]
Thank you so much for the detailed reply!

**Comment by Niviere** [comment]
so is it like, for example; I have 10 traits (eyes) all those 10 is equal to 100.
I can put whatever rarity number for each one as long as they all end up (added up) together as 100? (added up)

**Comment by bolshoytoster** [comment]
@Niviere yes, but they don't need to add up to 100.

**Comment by validusvell** [comment]
I have a trait group i only want to occur 10 percent of the time, can i apply the # to the folder? or do i have to create an "empty" trait in the group

**Comment by ermiasabera** [comment]
Good discussion,
what it looks like in multiple traits more than 2.


Example from 
Blue eye 
Brown eye 
Black eye 
And I want 100 NFTs with other traits. 

What if I want the maximum of 15% occurrence of blue eyes. And the other traits normal random. What would be the value of others ? 

Thank you

**Comment by bolshoytoster** [comment]
@ermiasabera you could have blue eyes as `#15` then have the rest be `85 / number_of_other_traits`.

**Comment by ermiasabera** [comment]
> @ermiasabera you could have blue eyes as `#15` then have the rest be `85 / number_of_other_traits`.

Thanks for your answer, 
I use hashlips no code tool/application, not the code. I will try to use this method there by changing # to $. Hopefully it works.

**Comment by bolshoytoster** [comment]
@ermiasabera
> I will try to use this method there by changing # to $. Hopefully it works.

I was talking about the `#number` in the image names, using a `$` won't work.

**Comment by ginofalcone** [comment]
I'm still confused in how to generate rarity for the new no code tool application. 

I'm used to doing everything in decimals .15 98.85 like described earlier. These examples here don't look like it's meant for the no coe application where the weight is with the $ 

if no $ sign is added, then by default 1 by default, correct? So how would I apply rarity to this? exact same way?

**Comment by bolshoytoster** [comment]
@ginofalcone could you provide a link to the 'new no code tool application' please; I can't find it.

**Comment by 0xinit** [comment]
so, let's say my background has 16 layers... 5 of them needs to appear for 12%, 5 of them appears to be 6% times, and other 6 layers have rarity anywhere from 0.5 to 5%. So how should I do it ? for layer with 5%, should it be # 5 and for  12% should it be # 12 ? or some other calculation?

**Comment by bolshoytoster** [comment]
@0xinit if all of your rarities add up to 100, then it will act as a percentage, but they will still work relatively regardless, i.e. # 12 will occur double as much as # 6.

**Comment by 0xinit** [comment]
so when we talk about "rarities" are we talking about just one trait like all the layers of "background" should add up to 100?

**Comment by bolshoytoster** [comment]
@0xinit yes, it is, but they only really need to add to 100 for simplicity.

**Comment by 0xinit** [comment]
@bolshoytoster ah that makes sense, because when people talk about rarities here, It thhought that all the layers which include background, base, eyes, ears etc.. all of them need to add to 100 as cumulative addition.

It makes sense tho that every folder will have their own 100 addition.

**Comment by 4x13** [comment]
seems like to make everything workout, we would need a blank transparent attribute in each layer folder, so if you wanted all of a certain attribute to be shown 10% of the time, lets say a Mask Attribute with 10 masks, we would assign a #1 to each mask and a #90 to the blank attribute to get to an easy 100 weight.

**Comment by Niviere** [comment]
And this for only for each map right?


On 30 May 2022, at 22:24, bolshoytoster ***@***.***> wrote:

﻿

@Niviere<https://github.com/Niviere> yes, but they don't need to add up to 100.

**Comment by bolshoytoster** [comment]
@Niviere
> map

Do you mean layer?
